sanskrit_parser/util/normalization.py

```diff
@@ -0,0 +1,51 @@
+'''
+Handles user-input and converts it into the format that internal modules expect
+and vice-versa
+
+@author:
+'''
+
+from __future__ import print_function, unicode_literals
+import re
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+def normalize(s):
+    """ Converts user-input into format expected by internal modules.
+    Input s is expected to be an SLP1 encoded string
+    """
+    def _dumpchars(s):
+        # Remove Unicode Zero-Width characters
+        s = re.sub("[\u200b\u200c\u200d]", "", s)
```

Comment:
Wondering if we could make all empty replacements a regex with character/numeric classes.

It may be bit more readable and efficient to use `string.maketrans` and build a static translation table. We use as many times required with `string.translate`.

Reply:
I replaced dumpchars with 
```python
    # Remove Unicode Zero-Width characters, punctuation and numeric characters
    s = re.sub("[\u200b\u200c\u200d,'-;().?!\"0123456789]", "", s)
    # Some bad visargas
    s = s.replace(':', 'H')
    # Replace line-breaks with spaces
    s = s.replace('\r\n', ' ').replace('\n', ' ')
```
Does that address the regex comment? or did you want to use something like \W to get rid of all of them at once? I think the reason to spell them out is that we don't want ":" to be removed in case someone entered it instead of a visarga.

The only replacement remaining is the line-breaks with spaces. My understanding is that maketrans needs inputs of equal length. Is \r\n considered one character or two ? In any case, I think the efficiency gain would be minimal.

Reply:
Yes, I meant a similar regex. Just played with `string.maketrans` little more. It is quite interesting, probably can reduce the complete logic to a single line. We can build a map for `":\n" -> "H "` and rest everything to just `None`. Probably an O(n), if we save the mapping after first invocation.

Something like...
```python
> python
Python 3.6.3 (default, Oct 24 2017, 14:48:20)                                                                                                                  
[GCC 7.2.0] on linux                                                                                                                                           
Type "help", "copyright", "credits" or "license" for more information.                                                                                         
>>> x = str.maketrans(":\n", "H ", "\u200b\u200c\u200d,'-;().?!\"0123456789\r")                                                                                
>>> x                                                                                                                                                          
{58: 72, 10: 32, 8203: None, 8204: None, 8205: None, 44: None, 39: None, 45: None, 59: None, 40: None, 41: None, 46: None, 63: None, 33: None, 34: None, 48: None, 49: None, 50: None, 51: None, 52: None, 53: None, 54: None, 55: None, 56: None, 57: None, 13: None}                                                         
>>> "a: some\r\nline 092wo-;rd".translate(x)                                                                                                                   
'aH some line word' 
```

Reply:
Unfortunately, on py2.7, this would seem require using https://docs.python.org/2/library/string.html#string.maketrans and the function signatures are quite different. The same can be achieved by compiling a regex pattern once, which is what c0845c5 does. Is that good?